kmsjwt.go

```diff
@@ -5,105 +5,97 @@ import (
-func New(api KMS, keyID string) *KMSJWT {
-	return &KMSJWT{api: api, keyID: keyID}
+// New retrieves the public key from KMS and returns a signer.
+func New(ctx context.Context, client KMS, keyID string) (KMSJWT, error) {
+	publicKey, err := getPublicKey(ctx, client, keyID)
+	if err != nil {
+		return KMSJWT{}, fmt.Errorf("kmsjwt new: %w", err)
+	}
+	return KMSJWT{client: client, keyID: keyID, publicKey: publicKey}, err
 }
 
-// NewWithPublicKey provides a KMS-based implementation of JWT signing method
-// with a pre-loaded public key.
-func NewWithPublicKey(api KMS, keyID string, publicKey *rsa.PublicKey) *KMSJWT {
-	return &KMSJWT{api: api, keyID: keyID, publicKey: publicKey}
+func getPublicKey(ctx context.Context, client KMS, keyID string) (*rsa.PublicKey, error) {
+	response, err := client.GetPublicKey(ctx, &kms.GetPublicKeyInput{KeyId: &keyID})
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve public key: %w", err)
+	}
+
+	publicKey, err := x509.ParsePKIXPublicKey(response.PublicKey)
+	if err != nil {
+		return nil, fmt.Errorf("could not parse public key: %w", err)
+	}
+
+	result, ok := publicKey.(*rsa.PublicKey)
+	if !ok {
+		return nil, fmt.Errorf("public key: cannot assert %T as %T", publicKey, result)
+	}
+
+	return result, nil
 }
 
-func (k *KMSJWT) Alg() string {
+// Alg returns the signing algorithm as defined in https://datatracker.ietf.org/doc/html/rfc7518#section-3.1.
+func (k KMSJWT) Alg() string {
 	return signingMethod.Alg()
 }
 
-func (k *KMSJWT) Sign(signingString string, key interface{}) (string, error) {
+// Sign signs the signingString with AWS KMS using the key ID stored on the object.
+// The key parameter expects a context.Context that is used for the network call to KMS.
+func (k KMSJWT) Sign(signingString string, key interface{}) (string, error) {
 	ctx, ok := key.(context.Context)
 	if !ok {
-		return "", jwt.ErrInvalidKeyType
+		return "", fmt.Errorf("kmsjwt sign: %w", jwt.ErrInvalidKeyType)
 	}
 
 	hash := signingMethod.Hash.New()
-	hash.Write([]byte(signingString))
+	_, _ = hash.Write([]byte(signingString))
 
-	out, err := k.api.Sign(ctx, &kms.SignInput{
+	out, err := k.client.Sign(ctx, &kms.SignInput{
 		KeyId:            aws.String(k.keyID),
 		Message:          hash.Sum(nil),
 		MessageType:      types.MessageTypeDigest,
-		SigningAlgorithm: types.SigningAlgorithmSpecRsassaPkcs1V15Sha512,
+		SigningAlgorithm: types.SigningAlgorithmSpecRsassaPssSha512,
 	})
-
-	if err != nil && errors.Is(err, context.Canceled) {
-		return "", err
-	} else if err != nil {
-		return "", jwt.ErrInvalidKey
-	}
-
-	return base64.RawURLEncoding.EncodeToString(out.Signature), nil
-}
-
-func (k *KMSJWT) Verify(signingString, stringSignature string, key interface{}) error {
-	ctx, ok := key.(context.Context)
-	if !ok {
-		return jwt.ErrInvalidKeyType
-	}
-
-	publicKey, err := k.getPublicKey(ctx)
 	if err != nil {
-		return err
+		return "", fmt.Errorf("kmsjwt signing with KMS: %w", err)
```

Comment:
Maybe just in order to preserve the original error

```suggestion
		return "", errors.Join(errors.New("kmsjwt signing with KMS"), err)
```

Reply:
See my previous answer about preserving the error. I prefer `fmt.Errorf("%w")` as that wrap. I don't want to have multiple errors with `errors.Join()`. I would use `errors.Join()` if I provided sentinel errors that the user can assert on with `errors.Is()`.